File: 1762-furthest-building-you-can-reach/1762-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int n = heights.size();
        priority_queue<int>pq;
        int sum = 0;
        for(int i=0; i<n-1; i++)
        {
            if(heights[i] >= heights[i + 1]) continue;
            
            int diff = heights[i + 1] - heights[i];

            pq.push(diff);
            sum += diff;
            if(sum > bricks)
            {
                if(ladders){
                    sum -= pq.top();
                    pq.pop();
                    ladders--;
                }
                else return i;
            }
        }
        return n - 1;
    }
};

File: 1762-furthest-building-you-can-reach/1762-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, nums: List[int], bricks: int, ladders: int) -> int:
        heap = []
        sum = 0

        n = len(nums)

        for i in range(n - 1):
            if nums[i + 1] - nums[i] <= 0:
                continue

            diff = nums[i + 1] - nums[i]

            heapq.heappush(heap, -1*diff)

            sum += diff

            if sum > bricks:
                if ladders > 0:
                    sum -= -1*heapq.heappop(heap)
                    ladders-=1
                else:
                    return i
        return n - 1